lib/dry/schema/messages/template.rb

```diff
@@ -0,0 +1,66 @@
+require 'dry/equalizer'
+
+require 'dry/schema/constants'
+
+module Dry
+  module Schema
+    module Messages
+      # Template wraps a string with interpolation tokens and defines evaluator function
+      # dynamically
+      #
+      # @api private
+      class Template
+        include Dry::Equalizer(:text)
+
+        TOKEN_REGEXP = /%{(\w+)}/
+
+        # !@attribute [r] text
+        #   @return [String]
+        attr_reader :text
+
+        # !@attribute [r] tokens
+        #   @return [Hash]
+        attr_reader :tokens
+
+        # !@attribute [r] evaluator
+        #   @return [Proc]
+        attr_reader :evaluator
+
+        # @api private
+        def self.[](input)
+          new(*parse(input))
+        end
+
+        # @api private
+        def self.parse(input)
+          tokens = input.scan(TOKEN_REGEXP).flatten(1).map(&:to_sym)
```

Comment:
`.flat_map(&:to_sym)` ;)

Reply:
This actually didn't work

Reply:
@solnic haha, right, that's because `flat_map` is actually `map_flat` 🤕

Reply:
Oops. Did not know this one. Will tweak it later :)

Reply:
your version is fine, though you can improve it with regex w/o capturing: `/(?<=%{)\w[\w\d]*(?=})/`, this will allow you to skip `flatten(1)`
```ruby
tokens = input.scan(TOKEN_REGEXP).map(&:to_sym)
```
also note that atm `TOKEN_REGEXP` allows digit-leading and empty tokens, both will break on `eval`

Reply:
@flash-gordon OK I will improve this